src/core/matcher.rkt

```diff
@@ -64,11 +64,29 @@
 
 (define (rewrite-once expr ctx #:rules rules #:root [root-loc '()])
   (define expr-repr (repr-of expr ctx))
-  (reap [sow]
-    (for ([rule rules] #:when (equal? expr-repr (rule-otype rule)))
-      (let* ([result (rule-apply rule expr)])
-        (when result
-          (sow (list (change rule root-loc (cdr result)))))))))
+  (define (match-otype? otype)
+    (or (equal? otype (representation-type expr-repr))
+        (equal? otype expr-repr)))
+
+  ;; we want rules over representations
+  (define-values (rules* canon-names) (expand-rules rules))
+  (define rule-apps (make-hash))
+
+  ;; actually match
+  (define changelists
+    (reap [sow]
+      (for ([rule rules*] #:when (match-otype? (rule-otype rule)))
+        (let* ([result (rule-apply rule expr)])
+          (when result
+            (define canon-name (hash-ref canon-names (rule-name rule)))
+            (hash-update! rule-apps canon-name (curry + 1) 1)
+            (sow (list (change rule root-loc (cdr result)))))))))
```

Comment:
Oof, for code that never runs, I gotta say this is pretty gnarly and kinda ugly. Do we actually want expanded rules in the stats? I feel like I'd be just as happy with the original rules showing up in the stats...

Reply:
I wanted the names in the timeline to be clean. It is a little gross, maybe I can clean it up.

Reply:
I think this code translates nice names to the ugly ones; perhaps we can just use the clean repr-generic names?

Reply:
Opposite, so I think I'll leave it.